typings/testing-library/index.d.ts

```diff
@@ -0,0 +1,7 @@
+/// <reference types="jest"/>
+
+declare namespace jest {
+	interface Matchers< R > {
+		toMatchStyleDiffSnapshot( expected: Element ): CustomMatcherResult;
+	}
```

Comment:
can we add an explanation for this declaration? it stands out in a number of ways to me that leave me confused:

 - Why does it exist at all?
 - What is the reason for the `R` type parameter?
 - Where doe we get the definition of `CustomMatcherResult`?

An explanatory comment could avoid a lot of confusion on those who are coming to this code and trying to understand it or modify it.

Reply:
Hey @dmsnell , thank you for taking a look at the code!

This declaration exists to offer types for the `toMatchStyleDiffSnapshot` custom matcher (defined in [`test/unit/config/matchers/to-match-style-diff-snapshot.js`](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/test/unit/config/matchers/to-match-style-diff-snapshot.js))

These are the definitions for the [`Matchers`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L690) and [`CustomMatcherResult`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L523-L526) interfaces.

After a few tries, I had landed on this format following [a few examples](https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300) on GitHub. I've now pushed a [change](https://github.com/WordPress/gutenberg/pull/39436/commits/d58fce70c42eb0e548c2c60cc2c5806cbe8ea106) to align the syntax for this declaration with [the one used in the `snapshot-diff` package](https://github.com/jest-community/snapshot-diff/blob/master/index.d.ts#L12-L24).

Reply:
It seems like `R` is the return type and probably comes from these returning `void|Promise<…>`. That's probably to allow both `expect().toBe()` and `expect().resolves.toBe()`.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d7d9c1d76dead09380e702d858b40432a29d8a48/types/jest/index.d.ts#L690

Did you consider putting these types in https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/test/unit/config/matchers/to-match-style-diff-snapshot.js? We might need to migrate that to TypeScript, but in my experience there's value in keeping the types close to what they're typing.

Reply:
thanks @ciampo! I appreciate the explanation but I found your comment here in GitHub clearer than the one in the code. Maybe that one in the code could use some language revision (read it out loud? have someone else read it and tell you what they think it means?)

otherwise, your comment here would make a great comment in the code. it doesn't really matter if I understand this while reviewing as much as it matters if the person who comes across this code in the future (could be yourself) gets the context for why it exists and what parts of it they can change without risking a problem.

```ts
/**
 * We use a custom matcher for the snapshot diff test which
 * takes into account project-specific behaviors and the
 * default matcher would invalidate those snapshots which
 * we consider valid.
 *
 * @see test/unit/config/matchers/to-match-style-diff-snapshot.js
 * @cite https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300
 */
```

Using the `R` type parameter makes it obvious why it's there. Is `T` necessary?  What happens if we omit it? Will the interface declaration still merge? If not, we can add a sentence like this to the comment.

```ts
/**
 * We use a custom matcher for the snapshot diff test which
 * takes into account project-specific behaviors and the
 * default matcher would invalidate those snapshots which
 * we consider valid.
 *
 * The `T` type parameter is necessary to match the interface
 * declaration from the `jest` types even though it isn't used
 * in this section.
 *
 * @see test/unit/config/matchers/to-match-style-diff-snapshot.js
 * @cite https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300
 */
```

Reply:
Hey @sirreal and @dmsnell, thank you for your feedback here! I'm going to try and answer all of the points raised.

> It seems like R is the return type and probably comes from these returning void|Promise<…>. That's probably to allow both expect().toBe() and expect().resolves.toBe().

That is correct, I did some further exploration and came to the same conclusion

> Did you consider putting these types in [`test/unit/config/matchers/to-match-style-diff-snapshot.js`](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/test/unit/config/matchers/to-match-style-diff-snapshot.js)? We might need to migrate that to TypeScript, but in my experience there's value in keeping the types close to what they're typing.

Things here get tricky. The signature of [the source `toMatchStyleDiffSnapshot` function](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/test/unit/config/matchers/to-match-style-diff-snapshot.js) which gets passed to `expect.extend()` is different from the signature of the function effectively invoked in the unit test.

I have spent some time looking into this and found [this article](https://haspar.us/notes/adding-jest-custom-matchers-in-typescript#going-too-far-with-types) which illustrates this difference in the signatures, and even proposes a way to compute the "user-facing" function type given the source matcher function types.

The above technique would likely require us to rewrite all of the [jest config to TypeScript](https://jestjs.io/docs/getting-started#using-typescript), something that @sarayourfriend suggested in https://github.com/WordPress/gutenberg/pull/39341#issuecomment-1065400890. This wouldn't be a trivial task (for the best results, [`ts-jest`](https://github.com/kulshekhar/ts-jest) should be used over Babel), but **I could open a new PR to explore this direction, if we think it's a better overall solution ?**

>  I appreciate the explanation but I found your comment here in GitHub clearer than the one in the code.
> ...
> it doesn't really matter if I understand this while reviewing as much as it matters if the person who comes across this code in the future (could be yourself) gets the context for why it exists and what parts of it they can change without risking a problem.

Thank you for the feedback on this one. It is true that this comment needs to be clear for the next person that comes across this code, but they should also explain what the `toMatchStyleDiffSnapshot` matcher does when used in a test.

I went ahead and updated the comment, taking inspiration from [the `toMatchSnapshot` comments](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d7d9c1d76dead09380e702d858b40432a29d8a48/types/jest/index.d.ts#L959-L962) and from the already existing docs that we have in [docs/contributors/code/testing-overview.md#best-practices](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/testing-overview.md#best-practices).

I also added the `@see` and `@cite` parts to the comment, as per your suggestion, although I didn't quite understand the part _"...and the default matcher would invalidate those snapshots which we consider valid."_. Could you explain further what you meant there?

I'd also be curious to hear @sarayourfriend 's opinion on what's the best description to add in the `toMatchStyleDiffSnapshot`'s docs.

> Using the R type parameter makes it obvious why it's there. Is T necessary? What happens if we omit it?

I removed the `T` parameter and it looks like everything still works as expected.

Reply:
On all points this sounds fair; thanks for taking the effort here to understand the types.

> Could you explain further what you meant there?

Your comment seems fine. Part of what makes mine awkward is that I understand the matcher type less than you do. I do have two quick nitpicks for the updated comment:

 - if we're using Markdown syntax, we might check if we can use the alternate link syntax since the inline `[title](URL)` format gets in the way of reading the code itself. if that works we should be able to follow the example below.
 - "associated to two different states" is probably smoother as "associated to _with_ different states" 

### alternate link syntax

```
/**
 * Similar to the [`toMatchDiffSnapshot` matcher], this custom matcher allows
 * to snapshot only the difference between the _styles_ associated to two
 * different states of a component.
 *
 * @see [to-match-style-diff-snapshot.js]
 * @see [Testing Overview docs]
 * @cite https://github.com/testing-library/react-testing-library/issues/36#issuecomment-440442300
 *
 * [`toMatchDiffSnapshot` matcher]: https://github.com/jest-community/snapshot-diff
 * [to-match-style-diff-snapshot.js]: https://github.com/WordPress/gutenberg/blob/trunk/test/unit/config/matchers/to-match-style-diff-snapshot.js
 * [Testing Overview docs]: https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/testing-overview.md#best-practices
 */
```

I'm not sure if our tools are capable of handling this syntax, so don't feel obliged to stick with it if it doesn't work out, but it would be nice to get rid of some of those long lines that are harder to parse when reading the source.

Reply:
Thank you for the feedback — I had forgotten about the shorthand reference syntax for markdown links!

I applied the requested changes in https://github.com/WordPress/gutenberg/pull/39436/commits/60e5015ff268269b2d923cc5f7349e4f104a86ee and https://github.com/WordPress/gutenberg/pull/39436/commits/2deef265a85d3f3c8d46b997118f282ea5cc5f37 .